graspy/match/faq.py

```diff
@@ -0,0 +1,250 @@
+    E. T. Harley, D. E. Fishkind, R. J. Vogelstein, and C. E. Priebe, “Fast
+    approximate quadratic programming for graph matching,” PLOS one, vol. 10, no. 4,
+    p. e0121002, 2015.
+
+    """
+
+    def __init__(
+        self,
+        n_init=1,
+        init_method="barycenter",
+        max_iter=30,
+        shuffle_input=True,
+        eps=0.1,
+    ):
+
+        if n_init > 0 and type(n_init) is int:
+            self.n_init = n_init
+        else:
+            msg = '"n_init" must be a positive integer'
+            raise TypeError(msg)
+        if init_method == "rand":
+            self.init_method = "rand"
+        elif init_method == "barycenter":
+            self.init_method = "barycenter"
+            self.n_init = 1
+        else:
+            msg = 'Invalid "init_method" parameter string'
+            raise ValueError(msg)
+        if max_iter > 0 and type(max_iter) is int:
+            self.max_iter = max_iter
+        else:
+            msg = '"max_iter" must be a positive integer'
+            raise TypeError(msg)
+        if type(shuffle_input) is bool:
+            self.shuffle_input = shuffle_input
+        else:
+            msg = '"shuffle_input" must be a boolean'
+            raise TypeError(msg)
+        if eps > 0 and type(eps) is float:
+            self.eps = eps
+        else:
+            msg = '"eps" must be a positive float'
+            raise TypeError(msg)
+
+    def fit(self, A, B):
+        """
+        Fits the model with two assigned adjacency matrices
+        
+        Parameters
+        ---------
+        A : 2d-array, square, positive
+            A square, positive adjacency matrix
+            
+        B : 2d-array, square, positive
+            A square, positive adjacency matrix
+
+        Returns
+        -------
+        
+        self : returns an instance of self
+        """
+        if A.shape[0] != B.shape[0]:
+            msg = "Matrix entries must be of equal size"
+            raise ValueError(msg)
+        elif A.shape[0] != A.shape[1] or B.shape[0] != B.shape[1]:
+            msg = "Matrix entries must be square"
+            raise ValueError(msg)
+        elif (A >= 0).all() == False or (B >= 0).all() == False:
+            msg = "Matrix entries must be greater than or equal to zero"
+            raise ValueError(msg)
+
+        n = A.shape[0]  # number of vertices in graphs
+
+        if self.shuffle_input:
+            node_shuffle_input = np.random.permutation(n)
+            P_shuffle_input = np.zeros((n, n))
+            # shuffle_input to avoid results from inputs that were already matched
+            P_shuffle_input[list(range(n)), node_shuffle_input] = 1
+            A = P_shuffle_input @ A @ np.transpose(P_shuffle_input)
+
+        At = np.transpose(A)  # A transpose
+        Bt = np.transpose(B)  # B transpose
+        score = math.inf
+        perm_inds = np.zeros(n)
+        grad_P = 1  # gradient of P
```

Comment:
@asaadeldin11  I've been playing with your code a bit - `grad_P` and `n_iter` both need to be initialized inside the `n_init` for loop. Otherwise, after the first iteration, `grad_P` is often something small, so the while loop doesn't get entered, so it is effectively only doing one random init. 

In general, I think this is why having tests would be really helpful now for the development process. Since your code keeps changing its hard to tell when something accidentally gets broken without tests. Also, the "proof" notebook you have should just import this code and use it, rather than define the whole class in the notebook, so we can tell how this class is actually doing.